File: LeetCode/143. Reorder List.java
/***  KNOWLEDGE IS POWER  ***/

class Solution {
    public void reorderList(ListNode head) {
        ArrayList<ListNode> dq = new ArrayList<ListNode>();
        ListNode p = head.next;
        while (p != null) {
            dq.add(p);
            p = p.next;
        }
        p = head;
        while (dq.size() > 0) {
            p.next = dq.get(dq.size() - 1);
            p = p.next;
            dq.remove(dq.size() - 1);
            if (dq.size() > 0) {
                p.next = dq.get(0);
                p = p.next;
                dq.remove(0);
            }
        }
        p.next = null;
    }
}

File: LeetCode/143. Reorder List.cpp
/***  KNOWLEDGE IS POWER  ***/

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode *p = head->next;
        vector<ListNode*> dq;
        while (p) {
            dq.push_back(p);
            p = p->next;
        }
        p = head;
        while (!dq.empty()) {
            p->next = dq.back();
            dq.pop_back();
            p = p->next;
            if (!dq.empty()) {
                p->next = dq.front();
                p = p->next;
                dq.erase(dq.begin());
            }
        }
        p->next = NULL;
    }
};
